### DISCUSSION TITLE
Making the text editor syntax-file, need help

### DISCUSSION THREAD

**Discussion by Alexey-T** [mdx-js/mdx#2301] [Q&A] [answered]
I want to make 'lexer' MDX for CudaText editor. 
I has JS, JSX, Markdown lexers.

I can clone the JSX syntax and I can add 'sub lexer' rules to include Markdown syntax for any ranges. But I don't know how to define Markdown ranges.
If Markdown block starts with some 'signature' and ends with some another 'signature' - question solved. But I see Markdown blocks don't have signatures. Then how can I detect MD ranges? Help.

**Comment by ChristianMurphy** [comment]
Welcome @Alexey-T!

If you are looking for inspiration on how to create syntax highlights for MDX, take a look at https://github.com/wooorm/markdown-tm-language

> I can clone the JSX syntax and I can add 'sub lexer' rules to include Markdown syntax for any ranges

MDX is JSX inside of Markdown, not the other way around.
Markdown is the starting lexer, and JSX is a sub lexer which starts when `{` or `<` are seen.

**Reply by ChristianMurphy** [reply]
Another fine grained note, there are a lot of Markdown flavors out there, MDX specifically extends CommonMark https://commonmark.org/

**Reply by Alexey-T** [reply]
JSX is a sublexer, got it. When JSX block is ended? e.g. this part: `<tag>text</tag>` - is this 2 JSX blocks and a text between them?

**Reply by ChristianMurphy** [reply]
yes, from a tokenizer perspective your example is tag, text, then tag again.
You can see a demo of the `markdown-tm-language` at https://wooorm.com/markdown-tm-language/ which shows the behavior.
Or if you are more interested in the AST structure, https://mdxjs.com/playground/ demonstrates that under the `mdast` tab.

**Reply by Alexey-T** [reply]
I see https://wooorm.com/markdown-tm-language/ with MDX checkmark. The document starts with a JS code - but how can I determine the start of this JS block from lexer perspective? and what is the end of that block?

**Reply by ChristianMurphy** [reply]
I'm not quite sure what you are asking @Alexey-T

Are you looking for a formal specification of MDX parsing? If so see https://github.com/micromark/mdx-state-machine (@wooorm may be able to comment on how up-to-date the spec is)

Or are you asking what the rules the demo is following in your example text? They are the following: https://github.com/wooorm/markdown-tm-language/blob/0d68d9239183956f1ddeec4b9de94654997407d2/source.mdx.js#L1593-L1710

Or something else entirely?

**Reply by Alexey-T** [reply]
I want to know about how my lexer can find JSX sub-blocks. you say, that I need these ranges

1. from `<` to `>`
2. from `{` to `}`

But the parser at https://wooorm.com/markdown-tm-language/ (when I check MDX checkbox) shows JSX also at the begin of the document! Is this additional 3rd rule? how does it work? maybe if finds JSX block from document-begin until the MD header `# ......` ?

**Reply by ChristianMurphy** [reply]
> 1. from `<` to `>`
> 2. from `{` to `}`

Roughly, there is some more nuance around the following:

```jsx
<Tag content={<AnotherTag number={1} />}>text</Tag>

{(() => {return "text"})}
```

`{}` and `<>` can be nested.

> shows JSX also at the begin of the document!

you can write a JSX tag early in the document, there's no special rule for it though.

> Is this additional 3rd rule?

There is no special rule for JSX at the start of the document.
But there are more rules/nuance on how to handle JSX in Markdown.

> how does it work? 

Again if you want the formal definition read https://github.com/micromark/mdx-state-machine

> maybe if finds JSX block from document-begin until the MD header `# ......` ?

I don't understand what you are asking.

Some examples of MDX syntax.
JSX can appear in a some markdown constructs, such as header.
```mdx
# <Tag />
```

is a valid tag in a markdown header.
and markdown can appear in JSX

```mdx
<Tag>*strong*</Tag>
```

will have strong text inside the JSX tag.

There is no requirement that a Markdown construct needs to happen before a JSX block.

**Reply by wooorm** [reply, marked as answer]
> JSX also at the begin of the document!

Do you me a jsx? Which is tag stuff. Or do you mean the import and export statements? If the latter, that's a third extension indeed.

You can find more on what mdx is. In all the preceding links, or on the website.
